src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,423 @@
+  private[this] def build(
+    podDefinition: PodDefinition,
+    offer: mesos.Offer,
+    newInstanceId: PathId => Instance.Id,
+    config: MarathonConf,
+    resourceMatch: ResourceMatcher.ResourceMatch
+  ): Some[(mesos.ExecutorInfo, mesos.TaskGroupInfo, Seq[Option[Int]])] = {
+    // TODO: probably set unique ID for each task
+    val instanceId = newInstanceId(podDefinition.id)
+
+    val allEndpoints = for {
+      container <- podDefinition.containers
+      endpoint <- container.endpoints
+    } yield endpoint
+
+    val portMappings = computePortMappings(allEndpoints, resourceMatch.hostPorts)
+
+    val executorInfo = computeExecutorInfo(podDefinition, resourceMatch.portsMatch, portMappings, instanceId)
+
+    val envPrefix: Option[String] = config.envVarsPrefix.get
+
+    val portsEnvVars = portEnvVars(allEndpoints, resourceMatch.hostPorts, envPrefix)
+
+    val taskGroup = mesos.TaskGroupInfo.newBuilder
+
+    podDefinition.containers
+      .map(computeTaskInfo(_, podDefinition, offer, instanceId, config, portsEnvVars))
+      .foreach(taskGroup.addTasks)
+
+    Some((executorInfo.build, taskGroup.build, resourceMatch.hostPorts))
+  }
+
+  // The resource match provides us with a list of host ports.
+  // Each port mapping corresponds to an item in that list.
+  // We use that list to swap the dynamic ports (ports set to 0) with the matched ones.
+  private[this] def computePortMappings(
+    endpoints: Seq[raml.Endpoint],
+    hostPorts: Seq[Option[Int]]): Seq[mesos.NetworkInfo.PortMapping] = {
+
+    endpoints.zip(hostPorts).collect {
+      case (mapping, Some(hostPort)) =>
+        if (mapping.containerPort.isEmpty || mapping.containerPort.get == 0) {
+          mesos.NetworkInfo.PortMapping.newBuilder
+            .setHostPort(hostPort)
+            .setContainerPort(hostPort).build
+        } else {
+          val portMapping = mesos.NetworkInfo.PortMapping.newBuilder
+            .setHostPort(hostPort)
+          mapping.containerPort.foreach(portMapping.setContainerPort)
+
+          // TODO(nfnt): set the protocol
+          portMapping.build
+        }
+    }
+  }
+
+  private[this] def computeTaskInfo(
+    container: raml.MesosContainer,
+    podDefinition: PodDefinition,
+    offer: mesos.Offer,
+    instanceId: Instance.Id,
+    config: MarathonConf,
+    portsEnvVars: Map[String, String]): mesos.TaskInfo.Builder = {
+    val builder = mesos.TaskInfo.newBuilder
+      .setName(container.name)
+      .setTaskId(mesos.TaskID.newBuilder.setValue(instanceId.idString))
+      .setSlaveId(offer.getSlaveId)
+
+    builder.addResources(scalarResource("cpus", container.resources.cpus))
+    builder.addResources(scalarResource("mem", container.resources.mem))
+    builder.addResources(scalarResource("disk", container.resources.disk))
+    builder.addResources(scalarResource("gpus", container.resources.gpus.toDouble))
+
+    val labels = podDefinition.labels ++ container.labels.map(_.values).getOrElse(Map.empty[String, String])
```

Comment:
someone else also thought that it would work this way. I always imagined that the pod definition labels would be placed into the executorInfo and that the container labels would be placed into the taskInfo. they would not be aggregated and then placed into taskInfo.